packages/tx-construction/src/tx-builder/TxBuilder.ts

```diff
@@ -287,6 +303,93 @@ export class GenericTxBuilder implements TxBuilder {
     }
   }
 
+  async #getOrCreateRewardAccounts(): Promise<RewardAccountWithPoolId[]> {
+    if (this.#requestedPortfolio) {
+      await util.createStakeKeys({
+        count: this.#requestedPortfolio.length,
+        keyAgent: this.#dependencies.keyAgent,
+        logger: contextLogger(this.#logger, 'getOrCreateRewardAccounts')
+      });
+    }
+
+    return this.#dependencies.txBuilderProviders.rewardAccounts();
+  }
+
+  async #delegatePortfolio(): Promise<void> {
+    if (!this.#requestedPortfolio) {
+      // Delegation using CIP17 portfolio was not requested
+      return;
+    }
+
+    // Create stake keys to match number of requested pools
+    const rewardAccounts = await this.#getOrCreateRewardAccounts();
+
+    // New poolIds will be allocated to un-delegated stake keys
+    const newPoolIds = this.#requestedPortfolio
+      .filter((cip17Pool) =>
+        rewardAccounts.every((rewardAccount) => rewardAccount.delegatee?.nextNextEpoch?.id !== cip17Pool.id)
+      )
+      .map(({ id }) => id)
+      .reverse();
+
+    this.#logger.debug('New poolIds requested in portfolio:', newPoolIds);
+
+    // Reward accounts which don't have the stake key registered or that were delegated but should not be anymore
+    const availableRewardAccounts = rewardAccounts
+      .filter(
+        (rewardAccount) =>
+          rewardAccount.keyStatus === Cardano.StakeKeyStatus.Unregistered ||
+          !rewardAccount.delegatee?.nextNextEpoch ||
+          this.#requestedPortfolio?.every(({ id }) => id !== rewardAccount.delegatee?.nextNextEpoch?.id)
+      )
+      .sort(GenericTxBuilder.#sortRewardAccountsDelegatedFirst)
+      .reverse(); // items will be popped from this array, so we want the most suitable at the end of the array
+
+    // Code below will pop items one by one (poolId)-(available stake key)
+    const certificates: Cardano.Certificate[] = [];
+    while (newPoolIds.length > 0 && availableRewardAccounts.length > 0) {
```

Comment:
Should I throw an internal error if `availableRewardAccounts.length < newPoolIds.length`?
The code before it creates the needed reward accounts, but the logic is a bit decoupled and a bug would result in some portfolio pools silently not delegated.

Reply:
Added here: [6a59b15](https://github.com/input-output-hk/cardano-js-sdk/pull/764/commits/6a59b15ee5301a1190840210c64c4f72ef0fb664).
I have tested the exception locally but haven't added a test due to mocking complications.